### ISSUE TITLE
Quote gets eaten as part of quadruple quote

### ISSUE THREAD

**Issue by MonoidMusician** [purescript/purescript#3959]
## Description

When quadruple doublequotes appear in a file, it should be parsed as the start of a triple-quoted block starting with a single doublequote character. Instead, it seems that the quote simply disappears and does not appear in the generated file.

## To Reproduce

```purescript
module Test where

s :: String
s = """"should start with a quote"""
```

Generated output:

```js
// Generated by purs version 0.13.3
"use strict";
var s = "should start with a quote";
module.exports = {
    s: s
};

```

## Expected behavior

The generated string should start with a quote: `var s = "\"should start with a quote"`.

## Additional context

I believe this is a regression from 0.12, since I included this in code that worked fine before upgrading to 0.13.

## PureScript version

0.13.3 and 0.14 (master branch)

**Comment by rhendric** [comment]
Can confirm this regressed in 0.13.0 (unsurprisingly, a bisect pegs the regression to 317dce2).

**Comment by rhendric** [comment]
I have some questions about the intended syntax specification for triple-quote literals.

Suppose that `""""x""" == "\"x"`, as it used to. Does `""""""x""" == "\"\"\"x"`, or is it a syntax error?

`"""x"""" == "x\""` currently. Should `"""x"""""" == "x\"\"\""`, as it currently does, or is it a syntax error?

Does `"""""" == ""`, or is that an unterminated string or some other syntax error? (It currently equals `"\"\"\"\"\""` due to what I assume is a mistaken `take` instead of a `drop` here:) https://github.com/purescript/purescript/blob/48634d05da1021ac1bcb4907461ae9b2d3b0e699/lib/purescript-cst/src/Language/PureScript/CST/Lexer.hs#L470-L472

Does `""""""" == "\""`, or is it an unterminated string or some other syntax error?

Basically, the most liberal spec I can imagine is something like, ‘A raw string literal is either a sequence of 6 or more double quotes, or a sequence that starts with 3–5 double quotes and is terminated by the next sequence of 3 or more double quotes. In either case, the value of the literal is equal to the raw syntax, with the first 3 and last 3 characters (all guaranteed to be double quotes) removed.’ (Note that this doesn't allow triplets of double quotes anywhere in the string except at the end.) But maybe a more conservative/less special-casey spec is desirable? In particular, maybe triplets of double quotes should be forbidden everywhere in the string, and `"""x""""""` should be made a syntax error?

cc @natefaubion

**Comment by garyb** [comment]
I came in here ready to give my answer to those examples thinking they were easily resolvable, but oh! It is a bit trickier than it first appears. :smile: Those are some very good questions.

**Comment by natefaubion** [comment]
The syntax specification is wrong. 

```
    string
      : '"' stringPart* '"'
      | '"""' .* '"""'
```

Or at least, it would require an alternative interpretation of `.*` 😆. This is too greedy, because otherwise you'd essentially never be able to have more than one raw string literal in a file, since the start of the first one would eat until the end of the last.

The intention is that a raw string literal is started by 3 quotes, and terminated by at least 3 quotes, where the leading quotes in the terminator are considered part of the string.

> Does """"""x""" == "\"\"\"x", or is it a syntax error?

It would be a syntax error, since it starts with 3 quotes, and ended by 3 quotes. That is a series of 6 quotes is the empty raw string literal. Otherwise you would have to forbid empty raw string literals (in lieu of `""`).

> Should """x"""""" == "x\"\"\"", as it currently does, or is it a syntax error?

It would be allowed as part of the terminator rule.

> Does """""" == "", or is that an unterminated string or some other syntax error? (It currently equals "\"\"\"\"\"" due to what I assume is a mistaken take instead of a drop here:) 

6 quotes is the empty raw string literal. It's a bug if that isn't the case.

> Does """"""" == "\"", or is it an unterminated string or some other syntax error?

It is a single quote considering the terminator rule.

> (Note that this doesn't allow triplets of double quotes anywhere in the string except at the end.)

Yes, this means that you cannot have triplets in the middle of a string, but I don't think this is new behavior.

I think the actual regex pattern for a raw string literal body is something like
```
([^"]+|"(?!"")|""(?!"))*
```
Where `(?!)` is negative lookahead.

**Comment by natefaubion** [comment]
I guess the full rule would be something like (in the comment's pseudo syntax):
```
'"""' ([^'"']+ | '"' (?!'""') | '""' (?! '"'))* '"'* '"""'
```

**Comment by rhendric** [comment]
> ```
> '"""' ([^'"']+ | '"' (?!'""') | '""' (?! '"'))* '"'* '"""'
> ```

Right, I *think* this is equivalent to the English spec I hypothesized.

After thinking about it more I think I'd like to advocate more strongly for the following variation:

```
'"""' ([^'"']+ | '"' (?!'""') | '""' (?! '"'))* '"'{0,2} '"""'
```

This changes ‘doesn't allow triplets of double quotes anywhere in the string except at the end’ to ‘doesn't allow triplets of double quotes anywhere in the string’, and so slightly reduces the expressiveness of raw string literals but makes the conceptual rule simpler and more symmetric. Another advantage is that it satisfies a sort of monotonicity property:

* If `'"""' w0 '"""'` is a valid raw string literal, where `w0` represents any sequence of characters, and if `w1` represents the result of inserting any `[^"]` character anywhere in `w0`, then `'"""' w1 '"""'` is also a valid raw string literal.

I think this property is an important part of my intuition about string literals, even though I've never expressed it quite this formally to myself. But with the unbounded terminator rule, the property can be violated: `"""a""""""` is valid but `"""a"""b"""` is not.

Thoughts? Does anyone feel strongly about losing the special ability to include more than two double quotes at the end of a raw string literal?

**Comment by natefaubion** [comment]
This is how it worked before 0.13 AFAIK. It means that you can’t write a raw string literal that ends in a quote (I think there’s an old issue about this?). I don’t think it’s common to have triple quotes in the middle of the string, so I don’t think it’s a big pain point. But not being able to end a raw string in a quote at all is inconvenient IMO. I know I’ve taken advantage of it.

**Comment by rhendric** [comment]
I don't mean to forbid `"""a""""` or `"""a"""""`. Just `"""a""""""` and up.

**Comment by natefaubion** [comment]
OK, I misunderstood some of your comment. Sorry for not reading more carefully. I think the idea of restricting it to a clear rule that it's never possible to have 3 subsequent quotes within a raw string literal is attractive, but I worry that it actually makes the rule more difficult to specify and parse. The current rule, while somewhat awkward to specify outside of prose (is there a better way than what I came up with above?), it's straightforward to parse, since you can just effectively `takeWhile '"'`. It's not really any more difficult than the old behavior, and eliminates an annoying edge case. Maybe that's too much in the weeds of implementations. I don't feel particularly strongly, I just think if we were to change it, we should try to change it as long as we can make it simpler in general. I'm happy to get more feedback on this.

**Comment by rhendric** [comment]
Re making the spec simpler, I'd propose:
```
'"""' '"'{0,2} ([^"]+ '"'{1,2})* [^"]* '"""'
```

(That's not an argument for or against our respective proposed rules, since your rule can be simplified similarly.)

Re implementation, it's not so difficult. If you add a new lexer monad operation `nextWhile' :: Int -> (Char -> Bool) -> Lexer Text` (which is similar to `nextWhile` but is bounded in the number of characters it consumes), the only other change necessary is replacing one `nextWhile` with `nextWhile' 7` and another with `nextWhile' 5` in the string lexing function (in addition to fixing the existing bugs in said function which prevent it from correctly handling your rule). I have a PR with tests ready to go if we get consensus that this is the way to go.

**Comment by natefaubion** [comment]
I don't have a problem adopting this rule. @garyb What do you think?

**Comment by garyb** [comment]
It seems reasonable to me :+1:

**Comment by hdgarrood** [comment]
This ought to be considered breaking, right? Since code which compiled under 0.13.x will still compile but with a different meaning?

**Comment by rhendric** [comment]
> This ought to be considered breaking, right? Since code which compiled under 0.13.x will still compile but with a different meaning?

#3961 has two distinct effects which would be considered breaking under the strictest possible interpretation:
 * Certain literals used to mean something highly unexpected, and now mean what a reasonable person would expect them to mean. (E.g., the initial report eating the first quote, or `""""""` meaning five quotes instead of the empty string.)
 * Certain literals used to compile and now are errors (any literal ending in three or more double quotes).

Whether or not these should actually be labeled as breaking depends on how strict you want to be, I guess. The first change corrects a bug on which I can't really imagine anyone intentionally relying (although it's true that someone could have written a raw literal with an extra quote at the start as a typo, and never noticed because of the bug; for them, upgrading PS will have the effect of revealing their own typo, but not in a way that would be caught at compile time). The second change will cause compilation errors for almost anyone who was relying on that feature, and it's an edge case anyway. If for some reason you're inclined to ‘pardon’ the most minor changes from breaking status, this would be a good candidate.

**Comment by garyb** [comment]
Maybe we need a category of "Breaking fix" :smile: - it's not unusual to have a bug fix that changes unintended behaviour in a way that corrects it, resulting in the potential for it to break for some people who relied on it.
